lib/transformer.js

```diff
@@ -0,0 +1,228 @@
+"use strict";
+
+const path = require("path");
+
+const Promise = require("bluebird");
+const fs = require("fs-extra");
+const webidl = require("webidl2");
+
+const Interface = require("./constructs/interface");
+const Dictionary = require("./constructs/dictionary");
+
+class Transformer {
+  constructor() {
+    this.options = { implSuffix: "" };
+    this.sources = [];
+    this._interfaces = null;
+    this._dictionaries = null;
+    this._customTypes = null;
+  }
+
+  addSource(idl, impl) {
+    if (typeof idl !== "string") {
+      throw new TypeError("idl path has to be a string");
+    }
+    if (typeof impl !== "string") {
+      throw new TypeError("impl path has to be a string");
+    }
+    this.sources.push({ idlPath: idl, impl });
+    return this;
+  }
+
+  *_collectSources() {
+    const stats = yield Promise.all(this.sources.map((src) => Promise.fromNode((next) => fs.stat(src.idlPath, next))));
+    const files = [];
+    for (let i = 0; i < stats.length; ++i) {
+      if (stats[i].isDirectory()) {
+        const folderContents = yield Promise.fromNode((next) => fs.readdir(this.sources[i].idlPath, next));
+        for (const file of folderContents) {
+          if (file.endsWith(".idl")) {
+            files.push({
+              idlPath: path.join(this.sources[i].idlPath, file),
+              impl: this.sources[i].impl
+            });
+          }
+        }
+      } else {
+        files.push({
+          idlPath: this.sources[i].idlPath,
+          impl: this.sources[i].impl
+        });
+      }
+    }
+    return files;
+  }
+
+  *_readFiles(files) {
+    return yield Promise.all(files.map((f) => Promise.props({
+      idlContent: Promise.fromNode((next) => fs.readFile(f.idlPath, { encoding: 'utf-8' }, next)),
+      impl: f.impl
+    })));
+  }
+
+  *_parse(outputDir, contents) {
+    const parsed = contents.map((content) => ({
+      idl: webidl.parse(content.idlContent),
+      impl: content.impl
+    }));
+
+    const interfaces = this._interfaces = {};
+    const dictionaries = this._dictionaries = {};
+    const customTypes = this._customTypes = new Set();
+
+    for (const file of parsed) {
+      for (const instruction of file.idl) {
+        switch (instruction.type) {
+          case "dictionary":
+            customTypes.add(instruction.name);
+            break;
+        }
+      }
+    }
+
+    for (const file of parsed) {
```

Comment:
I think we don't actually _need_ a 3 stage approach since nothing should actually kick of a generation during 1st and 2nd stage, so removed "stage 1" and commented the other two.